### PR TITLE
Download plugins from platform specific URLs

### PR DESCRIPTION
**Pull request by aoskotsky-amplify**

### CODE REVIEW

File: test/unit/test_file_download.py [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+class TestFileDownload(TestCase):
+    """Tests for file downloading utilities"""
+    @requests_mock.Mocker()
+    @patch('builtins.open', new_callable=mock_open)
+    @patch('os.path.isfile', MagicMock(return_value=True))
+    def test_file_download(self, mock_requests, open_mock):
+        """Test downloading a file"""
+        mock_requests.register_uri('GET', 'http://example.com', content=b'fake content')
+
+        download_file('http://example.com', '/tmp/plugins/foo')
```
Comment (aoskotsky-amplify) [inline test/unit/test_file_download.py:20]: Codacy Issue found: [Probable insecure usage of temp file/directory.](https://app.codacy.com/gh/amplify-education/terrawrap/pullRequest?prid=5656512)

**Comment by aoskotsky-amplify** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Coverage decreased per file
===========================
- terrawrap/models/graph.py  -5
         

Complexity increasing per file
==============================
- terrawrap/utils/file_download.py  9
- test/unit/test_file_download.py  2
         

Clones added
============
- test/unit/test_file_download.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/amplify-education/terrawrap/pullRequest?prid=5656512&bid=18517563)

**Comment by aoskotsky-amplify** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Coverage decreased per file
===========================
- terrawrap/models/graph.py  -5
         

Complexity increasing per file
==============================
- test/unit/test_file_download.py  2
- terrawrap/utils/file_download.py  9
         

Clones added
============
- test/unit/test_file_download.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/amplify-education/terrawrap/pullRequest?prid=5656512&bid=18517563)

File: test/unit/test_file_download.py (hunk 1 above)
Comment (aoskotsky-amplify) [inline test/unit/test_file_download.py:20]: Codacy Issue found: [Probable insecure usage of temp file/directory.](https://app.codacy.com/gh/amplify-education/terrawrap/pullRequest?prid=5656512)

**Comment by aoskotsky-amplify** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- terrawrap/utils/file_download.py  9
- test/unit/test_file_download.py  2
         

Clones added
============
- test/unit/test_file_download.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/amplify-education/terrawrap/pullRequest?prid=5656512&bid=18517563)

File: test/unit/test_file_download.py (hunk 1 above)
Comment (aoskotsky-amplify) [inline test/unit/test_file_download.py:20]: Codacy Issue found: [Probable insecure usage of temp file/directory.](https://app.codacy.com/gh/amplify-education/terrawrap/pullRequest?prid=5656512)

**Comment by aoskotsky-amplify** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- terrawrap/utils/file_download.py  9
- test/unit/test_file_download.py  2
         

Clones added
============
- test/unit/test_file_download.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/amplify-education/terrawrap/pullRequest?prid=5656512&bid=18517563)

File: test/unit/test_file_download.py (hunk 1 above)
Comment (aoskotsky-amplify) [inline test/unit/test_file_download.py:20]: Codacy Issue found: [Probable insecure usage of temp file/directory.](https://app.codacy.com/gh/amplify-education/terrawrap/pullRequest?prid=5656512)

**Comment by aoskotsky-amplify** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 6
           

Complexity increasing per file
==============================
- test/unit/test_plugin_download.py  2
- terrawrap/utils/plugin_download.py  9
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/amplify-education/terrawrap/pullRequest?prid=5656512&bid=18517563)

File: CHANGELOG.md [hunk 2]
```diff
@@ -0,0 +1,29 @@
+# Changelog
+
+All notable changes to this project will be documented in this file.
+
+The format follows [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
+and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).
+
+## [0.6.2] - 2020-06-08
```
Comment (aoskotsky-amplify) [inline CHANGELOG.md:8]: Codacy Issue found: [[no-shortcut-reference-link] Use the trailing `[]` on reference links](https://app.codacy.com/gh/amplify-education/terrawrap/pullRequest?prid=5656512)

File: CHANGELOG.md [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+### Added
+
+-   Added support for downloading plugins from platform specific URLs. 
+    For example on Mac, Terrawrap will try downloading from `<plugin_url>/Darwin/x86_64` first and fallback to 
+    downloading from `<plugin url>`. See README for more details.
+
```
Comment (aoskotsky-amplify) [inline CHANGELOG.md:15]: Codacy Issue found: [[list-item-spacing] Missing new line after list item](https://app.codacy.com/gh/amplify-education/terrawrap/pullRequest?prid=5656512)

File: CHANGELOG.md [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+    
+-   Added support for downloading from plugins from S3 using the AWS SDK
+
+## [0.6.1] - 2020-06-08
+
+### Changed
+
+-   Changed `backend_check` to ignore directories that don't contain `.tf` files
+
+## [0.6.0] - 2020-06-05
```
Comment (aoskotsky-amplify) [inline CHANGELOG.md:24]: Codacy Issue found: [[no-shortcut-reference-link] Use the trailing `[]` on reference links](https://app.codacy.com/gh/amplify-education/terrawrap/pullRequest?prid=5656512)

File: CHANGELOG.md [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+
+### Added
```
Comment (aoskotsky-amplify) [inline CHANGELOG.md:26]: Codacy Issue found: [[no-duplicate-headings] Do not use headings with similar content (10:1)](https://app.codacy.com/gh/amplify-education/terrawrap/pullRequest?prid=5656512)

File: CHANGELOG.md [hunk 6]
```diff
@@ -0,0 +1,29 @@
+# Changelog
```
Comment (aoskotsky-amplify) [inline CHANGELOG.md:1]: Codacy Issue found: [Command failed: git rev-parse --show-cdup
/bin/sh: git: not found
](https://app.codacy.com/gh/amplify-education/terrawrap/pullRequest?prid=5656512)

File: CHANGELOG.md [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+
+### Added
+
+-   Added support for downloading plugins from platform specific URLs. 
+    For example on Mac, Terrawrap will try downloading from `<plugin_url>/Darwin/x86_64` first and fallback to 
+    downloading from `<plugin url>`. See README for more details.
+    
+-   Added support for downloading from plugins from S3 using the AWS SDK
+
+## [0.6.1] - 2020-06-08
```
Comment (aoskotsky-amplify) [inline CHANGELOG.md:18]: Codacy Issue found: [[no-shortcut-reference-link] Use the trailing `[]` on reference links](https://app.codacy.com/gh/amplify-education/terrawrap/pullRequest?prid=5656512)

File: terrawrap/utils/plugin_download.py [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+        """
+        Download a file from either S3 or HTTP
+
+        :param url: URL where to download the file from
+        :param etag: etag value to use for caching
+        :return: File content and the file's etag. Will return None if the file is already cached
+        """
+        print('Downloading %s' % url)
+
+        if url.startswith('http://') or url.startswith('https://'):
```
Comment (jconstance-amplify) [inline terrawrap/utils/plugin_download.py:86]: It's probably cleaner to use https://docs.python.org/3/library/urllib.parse.html#url-parsing

Comment (jconstance-amplify) [reply terrawrap/utils/plugin_download.py:86]: https://github.com/amplify-education/amplify_aws_utils/blob/master/amplify_aws_utils/s3_uri.py#L21

File: CHANGELOG.md (hunk 6 above)
Comment (aoskotsky-amplify) [inline CHANGELOG.md:1]: Codacy Issue found: [Command failed: git rev-parse --show-cdup
/bin/sh: git: not found
](https://app.codacy.com/gh/amplify-education/terrawrap/pullRequest?prid=5656512)

**Comment by aoskotsky-amplify** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 6
           

Coverage increased per file
===========================
+ terrawrap/utils/graph.py  24
         

Coverage decreased per file
===========================
- terrawrap/models/graph.py  -5
         

Complexity increasing per file
==============================
- terrawrap/utils/plugin_download.py  9
- test/unit/test_plugin_download.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/amplify-education/terrawrap/pullRequest?prid=5656512&bid=18517563)

File: CHANGELOG.md (hunk 5 above)
Comment (aoskotsky-amplify) [inline CHANGELOG.md:26]: Codacy Issue found: [[no-duplicate-headings] Do not use headings with similar content (10:1)](https://app.codacy.com/gh/amplify-education/terrawrap/pullRequest?prid=5656512)

File: CHANGELOG.md (hunk 3 above)
Comment (aoskotsky-amplify) [inline CHANGELOG.md:15]: Codacy Issue found: [[list-item-spacing] Missing new line after list item](https://app.codacy.com/gh/amplify-education/terrawrap/pullRequest?prid=5656512)

File: CHANGELOG.md (hunk 7 above)
Comment (aoskotsky-amplify) [inline CHANGELOG.md:18]: Codacy Issue found: [[no-shortcut-reference-link] Use the trailing `[]` on reference links](https://app.codacy.com/gh/amplify-education/terrawrap/pullRequest?prid=5656512)

File: CHANGELOG.md (hunk 2 above)
Comment (aoskotsky-amplify) [inline CHANGELOG.md:8]: Codacy Issue found: [[no-shortcut-reference-link] Use the trailing `[]` on reference links](https://app.codacy.com/gh/amplify-education/terrawrap/pullRequest?prid=5656512)

File: CHANGELOG.md (hunk 4 above)
Comment (aoskotsky-amplify) [inline CHANGELOG.md:24]: Codacy Issue found: [[no-shortcut-reference-link] Use the trailing `[]` on reference links](https://app.codacy.com/gh/amplify-education/terrawrap/pullRequest?prid=5656512)

**Comment by aoskotsky-amplify** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 6
           

Complexity increasing per file
==============================
- test/unit/test_plugin_download.py  2
- terrawrap/utils/plugin_download.py  9
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/amplify-education/terrawrap/pullRequest?prid=5656512&bid=18517563)

File: CHANGELOG.md (hunk 4 above)
Comment (aoskotsky-amplify) [inline CHANGELOG.md:24]: Codacy Issue found: [[no-shortcut-reference-link] Use the trailing `[]` on reference links](https://app.codacy.com/gh/amplify-education/terrawrap/pullRequest?prid=5656512)

File: CHANGELOG.md (hunk 3 above)
Comment (aoskotsky-amplify) [inline CHANGELOG.md:15]: Codacy Issue found: [[list-item-spacing] Missing new line after list item](https://app.codacy.com/gh/amplify-education/terrawrap/pullRequest?prid=5656512)

File: CHANGELOG.md (hunk 7 above)
Comment (aoskotsky-amplify) [inline CHANGELOG.md:18]: Codacy Issue found: [[no-shortcut-reference-link] Use the trailing `[]` on reference links](https://app.codacy.com/gh/amplify-education/terrawrap/pullRequest?prid=5656512)

File: CHANGELOG.md (hunk 5 above)
Comment (aoskotsky-amplify) [inline CHANGELOG.md:26]: Codacy Issue found: [[no-duplicate-headings] Do not use headings with similar content (10:1)](https://app.codacy.com/gh/amplify-education/terrawrap/pullRequest?prid=5656512)

File: CHANGELOG.md (hunk 6 above)
Comment (aoskotsky-amplify) [inline CHANGELOG.md:1]: Codacy Issue found: [Command failed: git rev-parse --show-cdup
/bin/sh: git: not found
](https://app.codacy.com/gh/amplify-education/terrawrap/pullRequest?prid=5656512)

File: CHANGELOG.md (hunk 2 above)
Comment (aoskotsky-amplify) [inline CHANGELOG.md:8]: Codacy Issue found: [[no-shortcut-reference-link] Use the trailing `[]` on reference links](https://app.codacy.com/gh/amplify-education/terrawrap/pullRequest?prid=5656512)

File: terrawrap/utils/plugin_download.py [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+            if response.status_code == 304:
+                return None
+
+            return response.content, response.headers.get('etag')
+        except requests.HTTPError:
+            raise FileDownloadFailed()
+
+    def _get_s3_content(self, url: str, etag: Optional[str]) -> Optional[Tuple[bytes, Optional[str]]]:
+        """Download a file from S3 using the AWS SDK"""
+        match = re.match(r's3://([a-zA-Z0-9-.]+)/(.+)', url)
```
Comment (jconstance-amplify) [inline terrawrap/utils/plugin_download.py:118]: Nitpick: You could use urlparse to avoid this regex as well.

**Comment by aoskotsky-amplify** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 6
           

Complexity increasing per file
==============================
- test/unit/test_plugin_download.py  2
- terrawrap/utils/plugin_download.py  9
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/amplify-education/terrawrap/pullRequest?prid=5656512&bid=18517563)

File: CHANGELOG.md (hunk 5 above)
Comment (aoskotsky-amplify) [inline CHANGELOG.md:26]: Codacy Issue found: [[no-duplicate-headings] Do not use headings with similar content (10:1)](https://app.codacy.com/gh/amplify-education/terrawrap/pullRequest?prid=5656512)

File: CHANGELOG.md (hunk 4 above)
Comment (aoskotsky-amplify) [inline CHANGELOG.md:24]: Codacy Issue found: [[no-shortcut-reference-link] Use the trailing `[]` on reference links](https://app.codacy.com/gh/amplify-education/terrawrap/pullRequest?prid=5656512)

File: CHANGELOG.md (hunk 3 above)
Comment (aoskotsky-amplify) [inline CHANGELOG.md:15]: Codacy Issue found: [[list-item-spacing] Missing new line after list item](https://app.codacy.com/gh/amplify-education/terrawrap/pullRequest?prid=5656512)

File: CHANGELOG.md (hunk 6 above)
Comment (aoskotsky-amplify) [inline CHANGELOG.md:1]: Codacy Issue found: [Command failed: git rev-parse --show-cdup
/bin/sh: git: not found
](https://app.codacy.com/gh/amplify-education/terrawrap/pullRequest?prid=5656512)

File: CHANGELOG.md (hunk 7 above)
Comment (aoskotsky-amplify) [inline CHANGELOG.md:18]: Codacy Issue found: [[no-shortcut-reference-link] Use the trailing `[]` on reference links](https://app.codacy.com/gh/amplify-education/terrawrap/pullRequest?prid=5656512)

File: CHANGELOG.md (hunk 2 above)
Comment (aoskotsky-amplify) [inline CHANGELOG.md:8]: Codacy Issue found: [[no-shortcut-reference-link] Use the trailing `[]` on reference links](https://app.codacy.com/gh/amplify-education/terrawrap/pullRequest?prid=5656512)

**Comment by aoskotsky-amplify** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 6
           

Coverage increased per file
===========================
+ terrawrap/utils/graph.py  24
         

Coverage decreased per file
===========================
- terrawrap/models/graph.py  -5
         

Complexity increasing per file
==============================
- terrawrap/utils/plugin_download.py  9
- test/unit/test_plugin_download.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/amplify-education/terrawrap/pullRequest?prid=5656512&bid=18517563)

File: CHANGELOG.md (hunk 7 above)
Comment (aoskotsky-amplify) [inline CHANGELOG.md:18]: Codacy Issue found: [[no-shortcut-reference-link] Use the trailing `[]` on reference links](https://app.codacy.com/gh/amplify-education/terrawrap/pullRequest?prid=5656512)

File: CHANGELOG.md (hunk 2 above)
Comment (aoskotsky-amplify) [inline CHANGELOG.md:8]: Codacy Issue found: [[no-shortcut-reference-link] Use the trailing `[]` on reference links](https://app.codacy.com/gh/amplify-education/terrawrap/pullRequest?prid=5656512)

File: CHANGELOG.md (hunk 6 above)
Comment (aoskotsky-amplify) [inline CHANGELOG.md:1]: Codacy Issue found: [Command failed: git rev-parse --show-cdup
/bin/sh: git: not found
](https://app.codacy.com/gh/amplify-education/terrawrap/pullRequest?prid=5656512)

File: CHANGELOG.md (hunk 4 above)
Comment (aoskotsky-amplify) [inline CHANGELOG.md:24]: Codacy Issue found: [[no-shortcut-reference-link] Use the trailing `[]` on reference links](https://app.codacy.com/gh/amplify-education/terrawrap/pullRequest?prid=5656512)

File: CHANGELOG.md (hunk 5 above)
Comment (aoskotsky-amplify) [inline CHANGELOG.md:26]: Codacy Issue found: [[no-duplicate-headings] Do not use headings with similar content (10:1)](https://app.codacy.com/gh/amplify-education/terrawrap/pullRequest?prid=5656512)

File: CHANGELOG.md (hunk 3 above)
Comment (aoskotsky-amplify) [inline CHANGELOG.md:15]: Codacy Issue found: [[list-item-spacing] Missing new line after list item](https://app.codacy.com/gh/amplify-education/terrawrap/pullRequest?prid=5656512)

**Pull request merged**